blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+
+With that caveat in mind. Here is some real benchmark numbers of overhead for
+the julia test suite. The tests suite mostly uses message-passing based
+parallelism for running multiple tests in parallel. Only the `threads` test
+incurs the shared memory overhead penalty. Here are the measurements:
+
+[data]
+
+The total compressed size of a trace for a full run of the test suite (about
+30 minutes on 10 cores) is about three gigabytes.
+
+# Hardware and Software limitations
+
+As was already mentioned, `rr` currently only works on Linux. However, there
+are more restrictions. At the moment only x86 chips with Intel microarchitecture
+are supported. `rr` relies on hardware performance counters to establish a
+precise and accurate notion of time for asynchronous events. If these hardware
+counters are not available or not precise enough, recording using `rr` will not
+work. We and others have investigated whether it would be possible to port rr
+to other architectures. To the best of my knowledge, the current thinking on
+this is as follows:
+
+- For x86_64 with AMD Ryzen microarchitecture, it seems potentially possible, but the
+    AMD performance counters are less accurate than those found on Intel chips.
+    It may be possible to compensate for this in software, but the in-accuracies
+    are ill-understood. Help in investigation is welcome ([issue](https://github.com/mozilla/rr/issues/2034)).
+
+- For Aarch64, ll/sc-based atomics introduce additional problems, since various
+    external factors (interrupts, descheduling, etc.) can be observed as sc aborts.
+    With proper hardware support, this would be possible to work around, but such
+    hardware support appeared to be too in-accurate on older generation Aarch64 chips
+    and have been removed on recent generation ones. That said, Aarch64
+    microarchitectures are varied and as evidenced by the x86_64 experience, these
+    things can vary quite a bit by microarchitecture. Investigation of the hardware
+    capabilities of various Aarch64 microarchitectures would be very welcome
+    ([issue](https://github.com/mozilla/rr/issues/1373)).
+
+- We have looked into rr on POWER9, which is found in several large supercomputers
+    that run julia jobs. POWER9 has similar issues to Aarch64 since it is also an
+    ll/sc architecture, as well as supporting hardware transactional memory. In theory,
+    we believe the hardware exists to compensate for these issues, but preliminary
+    investigation suggests the hardware is not accurate enough to support `rr`. That
+    said, I understand there's active discussions with IBM to determine whether this
+    analysis is correct.
+
+One additional point of complication, is that it is quite easy for hypervisors
+(such as those used by cloud vendors) to interfere with the CPU capabilities
+required by `rr`. My understanding is that `rr` works fine on the latest generation
+of AWS machines, but not on GCP or Azure.
```

Comment:
You can confidently state that rr does not work on GCP or Azure but does work on AWS instances where hardware performance counters are enabled. (E.g. AWS C-instances where the VM occupies a whole socket, e.g. c5.9xlarge and bigger ... though those details don't matter for this post.)

Reply:
Done.